### PR TITLE
MultiChannel Scheduler

### PR DESCRIPTION
**Pull request by gjesse**

### CODE REVIEW

**Comment by davewingate** [comment]
Just trying to orient my self for CR ... I assume the intent here is to provide fair scheduling for jobs that correspond to, say, different social channels?

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelScheduler.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+ * dimension and attempts to maintain fairness among jobs in a particular by round-robining
+ * through all channels.
+ *
+ * Includes support for methods from {@link StateDedupedJobScheduler} and it's abstract hierarchy,
+ * however does not extend from that tree because some methods require the channel parameter
+ *
+ * it may prove useful to create a MultiChannelScheduler.ForChannel that extends from the {@link AbstractDedupJobScheduler} hierarchy
+ *
+ * // TODO This doesn't yet incorporate concepts from https://github.com/lithiumtech/rdbi/commit/6bbf2eeb49b87b71655f24fa9b797300b37b6797, that will be tackled separately
+ */
```
Comment (davewingate) [inline rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelScheduler.java]: Suggest adding example usage for scheduling and then reserving jobs.  For example, after I've reserved a job and finished it, do I need to ack completion back to this api?  Also, if I reserve a job and fail to process it b/c my jvm is restarted, what recourse do I have in the api  to let some other jvm pick up the scheduled job?

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelScheduler.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        this.clock = clock;
+    }
+
+    public MultiChannelScheduler(RDBI rdbi, String redisPrefixKey) {
+        this(rdbi, redisPrefixKey, System::currentTimeMillis);
+    }
+
+    /**
+     * see {@link AbstractDedupJobScheduler#schedule(String, String, int)}
+     *
```
Comment (davewingate) [inline rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelScheduler.java]: suggest adding javadoc explaining what the "channel" and "tube" concepts mean.

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelScheduler.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+     * @return true if the job was scheduled.
+     *         false indicates the job already exists in the ready queue.
+     */
+    public boolean schedule(String channel, String tube, final String job, final int runInMillis) {
```
Comment (davewingate) [inline rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelScheduler.java]: If I'm understanding correctly, there's a tube and that tube has multiple channels within that tube.  If that's right, I'd prefer to to not have to repeat the tube param through all the apis methods in this class.  Instead, how about 
```
MultiChannelScheduler scheduler = MultiChannelScheduler.builder()
                                                                                                   .rdbi()
                                                                                                   .redisPrefixKey()
                                                                                                   .channel()
                                                                                                   .clock()
                                                                                                   .build();

scheduler.schedule(String tube, String job, int runInMillis)

```

That way there's less of a chance that api consumers will inadvertently swap their channel and tube strings.

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelSchedulerDAO.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+import com.lithium.dbi.rdbi.BindArg;
+import com.lithium.dbi.rdbi.BindKey;
+import com.lithium.dbi.rdbi.Mapper;
+import com.lithium.dbi.rdbi.Query;
+
+import java.util.List;
+
+/**
+ *
+ * This is a pure round robin scheduler. If the next channel up doesn't have data it returns immediately
```
Comment (davewingate) [inline rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelSchedulerDAO.java]: I'm not sure I understand what you mean here.  If channels A & B have no scheduled jobs, but channel C has dozens of scheduled jobs, what happens when I call ```reserveMulti```?  Do I have to call multiple times, working my way through the empty queues before I get the scheduled jobs in channel C?

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelScheduler.java (hunk 1 above)
Comment (gjesse) [reply rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelScheduler.java]: sure, i can expand on it some.. there's a fair bit of implicit knowledge required based on how we use the other schedulers. 

as for handling the failure scenarios you described there's nothing at this level that dictates you it's done. The only recourse available is to look for expired running jobs, and then decide if they should be retried or not.

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelScheduler.java (hunk 3 above)
Comment (gjesse) [reply rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelScheduler.java]: yeah that's a good point, but i'm not sure about the use case from, say omega, which is going to be constantly scheduling jobs for different tubes + channels, is the overhead of maintaining a cache of schedulers per tube (or channel (or both)) worth it?

alternately i could create a type for Channel that's a simple string wrapper.. i'll consider it some more.

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelSchedulerDAO.java (hunk 4 above)
Comment (gjesse) [reply rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelSchedulerDAO.java]: no, you should only get C at that point

**Comment by gjesse** [comment]
overall, yes the intent is to give fairness to different entities across tubes, and the channels might correspond to different companies. 

for example current usage in omega, a tube == job type == something like FACEBOOK_MESSENGER or similar. jobs submitted to that tube are read fifo more or less, so an abundance of jobs from a single source end up affecting everyone - the goal is that they only affect themselves in these cases.

File: rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelSchedulerDAO.java (hunk 4 above)
Comment (gjesse) [reply rdbi-recipes/src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/MultiChannelSchedulerDAO.java]: actually this is a misleading comment from the original POC that Phu did - it no longer applies.

**Pull request merged**